A34-weighted-round-robin.md

```diff
@@ -0,0 +1,111 @@
+- `weighted_round_robin` updates the entries in EDF priority queue on any change of subchannel `ConnnectiveState` or endpoint `load_balancing_weight`.
+- If all endpoints have the same `load_balancing_weight`, `EDF` picker degenerates to `round_robin` picker. It's easier to reason and consistent with envoy.
+- Endpoints do not have `load_balancing_weight` assigned are discarded.
+- `weighted_round_robin` should always be updated to the lastest `ClusterLoadAssignment`. It's xDS server's responsibility to maintain consistency.
+
+#### EDF picker interface
+```
+/*
+Add new SubChannel to EDF picker.
+
+weighted_round_robin sees a new SubChannel in the most recent ClusterLoadAssignment or ConnectivityState of SubChannel is READY again.
+*/
+Add(SubChannel, weight)
+
+/*
+Remove Subchannel from EDF picker.
+
+Subchannel is removed from the most recent ClusterLoadAssignment comparing to last ClusterLoadAssignment or
+ConnectivityState of SubChannel becomes not READY.
+*/
+Remove(SubChannel)
+
+/*
+Update the weight of SubChannel
+
+There is weight change on Subchannel in the most recent ClusterLoadAssignment comparing to last ClusterLoadAssignment
+*/
+Update(SubChannel, newWeight)
+
+/*
+`Pick` picks the EdfEntry on top of the queue, returns the underlying SubChannel and updates deadline of the EdfEntry.
+*/
+PickResult Pick(PickArgs args)
+```
+
+### On new `ClusterLoadAssignment`
+
+#### endpoints have weight change
+- Update EDF priority queue by updating entries have weight change.
+- Different from [Envoy](https://github.com/envoyproxy/envoy/blob/51551ae944c642e6fc61563cbea8653087e70f1f/source/common/upstream/load_balancer_impl.cc#L733-L737), we'd like to udpate EDF priority queue so that new weights applied immediately even endpoints list is not changed.
+
+#### endpoints list change
+- Udpate EDF priority queue by adding/removing entries to reflect the change immediately.
+
+### On endpoint ConnectiveState Update
+- Udpate EDF priority queue by adding/removing entries to reflect the change immediately.
+
+## Rationale
+
+Several applications can be built upon this feature, e.g. utilization load balancing, blackhole erroring endpoints, load testing,... etc.
+
+The reason to refresh EDF picker even there is only weight change on some endpoints which is different from envoy is because we'd like real time traffic shift for use cases like load testing, blackhole erroring endpoints.
+
+The reasons to introduce a new algorithm instead of using existing `weighted_target` policy are
+- [EDF](https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) maintains FIFO order for endpoints with same weight which is easier to reason.
+- We want to be consistent with the behavior of Envoy.
+
+## Implementation
+
+N/A
+
+## Open issues (if applicable)
+
+- Replace `round_robin` with `weighted_round_robin`?
+- Following first issue, if not, do we want to use `weighted_round_robin` as default lb_policy for eDS response?
```

Comment:
Since this is more in-line with Envoy's behavior, I think we can use this policy when the CDS response sets the LB policy to `ROUND_ROBIN` (which is the only value we accept today, although that may change in the not-too-distant future).